File: CPP/ByRound/1739/1739_E_Cleaning_Robot.cpp
// Problem: CF 1739 E - Cleaning Robot
// https://codeforces.com/contest/1739/problem/E

/*
 * Problem: Cleaning Robot
 * 
 * Algorithm/Technique: Dynamic Programming with Memoization
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * 
 * The problem involves a robot cleaning a hallway represented as a 2xN grid.
 * The robot starts at (1,1) and always moves to the closest dirty cell.
 * If multiple closest cells exist, the robot malfunctions.
 * 
 * We use dynamic programming to determine the maximum number of dirty cells
 * that can be left such that the robot never faces a tie in choosing the
 * closest dirty cell.
 * 
 * The DP state dp[p][q][r] represents the maximum number of dirty cells we
 * can leave in the subproblem starting at column p, with current row q,
 * and flag r indicating if we've just moved from the opposite row (r=1) or not (r=0).
 * 
 * Key idea:
 * - We track the maximum number of dirty cells that can be cleaned,
 *   ensuring that at each step, there's a unique closest dirty cell.
 * - The algorithm explores both possibilities: 
 *   1. Move to the same row (no need to track previous row)
 *   2. Move to the opposite row (if there's a dirty cell there and we're not restricted)
 */

#include <stdio.h>
#include <string.h>
#include <algorithm>
#include <iostream>
#include <string>

using namespace std;
int dp[200005][2][2];
string s[2];
int n;
int dpp(int p, int q, int r) {
  int &ans = dp[p][q][r];
  if (p == n)
    ans = 0;
  if (~ans)
    return ans;
  int id = s[q][p] - '0';
  ans = id + dpp(p + 1, q, 0);
  if (s[1 - q][p] == '1' && !r)
    ans = max(id + 1 + dpp(p + 1, 1 - q, 1), ans);
  return ans;
}
int main() {
  memset(dp, -1, sizeof(dp));
  scanf("%d", &n), cin >> s[0] >> s[1];
  printf("%d", dpp(0, 0, 0));
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/E/1739_E_Cleaning_Robot.py
# Problem: CF 1739 E - Cleaning Robot
# https://codeforces.com/contest/1739/problem/E

# Flowerbox for code purpose and complexity
"""
Code Purpose:
This code solves the "Cleaning Robot" problem where a robot moves through a 2xN grid to clean dirty cells.
The robot always moves to the closest dirty cell, but if there are multiple closest ones, it malfunctions.
We must determine the maximum number of dirty cells that can be left for the robot to clean without malfunctioning.

Approach:
- Dynamic Programming with state f[i][j] representing max number of dirty cells cleaned
  when the robot is at row j, column i.
- Two transitions are possible:
  1. Move right to next column in same row.
  2. Move to the other row at the same column, if there's a dirty cell there (with additional cost).

Time Complexity: O(n)
Space Complexity: O(n)
"""

import sys

inf = float("inf")
mod = 998244353
input = lambda: sys.stdin.readline().rstrip()
inpnm = lambda: map(int, input().split())
inparr = lambda: [int(i) for i in input().split()]
inpint = lambda: int(input())


n = inpint()
s = []
for i in range(2):
    s.append([i for i in list(input())] + ["0", "0"])

f = [[-inf] * (2) for _ in range(n + 2)]
f[0][0] = 0
for i in range(n):
    for j in range(2):
        f[i + 1][j] = max(f[i + 1][j], f[i][j] + int(s[j][i + 1]))
        if s[j ^ 1][i] == "1":
            f[min(i + 2, n)][j ^ 1] = max(
                f[min(i + 2, n)][j ^ 1],
                f[i][j] + 1 + int(s[j ^ 1][i + 1]) + int(s[j ^ 1][i + 2]),
            )
print(max(f[n][0], f[n][1]))


# https://github.com/VaHiX/CodeForces/